### ISSUE TITLE
Some questions about model and fine-tuning

### ISSUE THREAD

**Issue by linchuming** [sniklaus/softmax-splatting#5]
Nice work!
But, some questions trouble me after I read the paper:
1. GridNet has only one input and the outputs of feature pyramid extractor are three features with different space sizes. How to combine the outputs to the GridNet?
2. How to fine-tune the PWC-Net with your model? Do I need add additional loss?
3. Feature pyramid will generate features of three scales. To splat the three features to the target frame, we need three important metrics corresponding to three features.  According to your paper, we can produce the important metric of original scale.  So is the low scale metric obtained by downsampling the original scale metric?

**Comment by sniklaus** [comment]
Thank you for your interest in our paper!

1. We have multiple inputs to the GridNet, one per level as shown in the illustration in the figure below. For your convenience, I also added the number of output features for each block.

![Screenshot_2020-04-14_09-29-59](https://user-images.githubusercontent.com/1238034/79250294-49836080-7e33-11ea-9b76-643da82e1b09.png)

2. We initially take an off-the-shelf PWC-Net and only optimize the parameters of our frame synthesis network and our feature pyramid extractor (and the scale parameters for each softmax splatting operator, to be precise). When fine-tuning PWC-Net, we do not change our loss, all we do is additionally backpropagating though PWC-Net and updating its weights too.
3. You can consider each softmax splatting as its own independent operation, my apologies that the figure in the paper shows them as one block. For the softmax splatting at the first level, we compute the photometric consistency for the importance map from the input images and the features in the first level of the feature pyramid. For the second level, we compute the photometric consistency for the importance map from the features in the second level of the feature pyramid (and so on). You could also take importance metric calculated at the first level and downscale it to use it for the second and third level. It has been a while since I did this experiment, but the interpolation quality was equally well if I remember correctly.

**Comment by linchuming** [comment]
@sniklaus Thanks for your detailed response!
I tried to train the interpolation model with PWC-Net without adding additional loss, but the PWC-Net is unable to converge. Should I adjust the learning rate for PWC-Net?

Another problem is that softmax splatting includes a small U-Net with three levels to refine the important map . So does each softmax splatting module has a independent U-Net?

**Comment by sniklaus** [comment]
Would you mind elaborating on your current state of progress? Have you implemented and trained the frame synthesis network as well as the feature pyramid extractor and you know want to fine-tune PWC-Net after successfully having trained the rest of the pipeline already?

I just checked the importance metric calculation again and noticed that we only compute it for the highest level and then downsample it to obtain the importance metric for the lower levels. I accordingly updated my response above. My apologies for having it mixed up, it has been almost two years since I wrote the code for this project.

Regarding the U-Net to fine-tune the importance metric, please see my response in #4 which you may find useful. In there, I recommend to first start trying without using this component since the effect of fine-tuning the importance metric is minor anyway.

**Comment by linchuming** [comment]
@sniklaus Thanks for your reponse again!
I tired to reimplement your softspalting approach. I use some configs as follows:
1. feature pyramid extractor and GridNet as your paper.
2. compute important metric by images and downsample it for lower levels.
3. use the Lap loss:
`class LaplacianLoss(nn.Module):
    def __init__(self):
        super(LaplacianLoss, self).__init__()

        self.criterion = nn.L1Loss()
        self.lap = LaplacianPyramid()

    def forward(self, x, y):
        x_lap, y_lap = self.lap(x), self.lap(y)
        return sum(self.criterion(a, b) for a, b in zip(x_lap, y_lap))

`

**Comment by sniklaus** [comment]
Looks like you got pretty good results already, congrats!

3. You could try weighting the levels differently, for example giving the coarser levels in the pyramid more weight. Your mileage may vary, there is definitely room for improvement though.
4. Notice that we used a PWC-Net pre-trained on FlyingChairs, which is not an official model. So you may get worse or better results than us.
7. You might want to try a few to see what works best for you. Lowering the learning rate towards the end of your training may be worth a shot, too.

Good luck!

**Comment by Hsveh** [comment]
I used the same method as @linchuming  and got a similar result (PSNR:34.95). Therefore, according to your suggestion, I added different weights to each layer in the pyramid. Now the loss function I use is as follows:
`class LaplacianPyramid(nn.Module):
    def __init__(self, max_level=5):
        super(LaplacianPyramid, self).__init__()
        self.gaussian_conv = GaussianConv()
        self.max_level = max_level

    def forward(self, X):
        t_pyr = []
        current = X
        for level in range(self.max_level):
            t_guass = self.gaussian_conv(current)
            t_diff = current - t_guass
            t_pyr.append(t_diff)
            current = F.avg_pool2d(t_guass, 2)
        t_pyr.append(current)

        return t_pyr

class LaplacianLoss(nn.Module):
    def __init__(self):
        super(LaplacianLoss, self).__init__()

        self.criterion = nn.L1Loss()
        self.lap = LaplacianPyramid()

    def forward(self, x, y):
        x_lap, y_lap = self.lap(x), self.lap(y)
        weights= [1, 2, 4, 8, 16, 32]
        return sum(weights[i] * self.criterion(a, b) for a, b in enumerate(zip(x_lap, y_lap)))`
But I got a worse result(PSNR: 34.27). Could you please tell me how to change my loss function.
